utils/docs/gnet_hier_verilog.1
.TH gnet_hier_verilog.sh 1 "September 25th, 2013" "gEDA Project" 1.8.2.20130925

.SH NAME
gnet_hier_verilog.sh \- generate a non-flattened hierarchical Verilog netlist.

.SH SYNOPSIS
.B gnet_hier_verilog.sh
.I filename.sch

.SH DESCRIPTION
.B gnet_hier_verilog.sh
is a simple script to produce a
hierarchical verilog netlist in a single file.
It gathers hierarchical information from a list of
unique symbols/schematics originating from the top level
schematic all the way down to the lowest level of the
design hierarchy. It then successively invokes the
existing gEDA verilog netlister to produce each single
level netlists, and concatinates all the unique
module netlists into one single hierarchical netlist
file.
.PP
Currently,
.B gnet_hier_verilog
assumes that one or more hierarchical symbol
can be represented by a single schematic file. If needed,
feature for mutiple schematic files mapped to a single symbol
can be easily added. In that case, multiple source attrib
are used in that symbol.

.SH AUTHOR
Paul Tan

.SH SEE ALSO
.BR gschem (1),
.BR gnetlist (1)

.SH COPYRIGHT
.nf
Copyright \(co 1999-2011 gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2014 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 1.9.2
=================================
* Visibility and attributes of several VHDL symbols have been fixed.

* Two obsolete scripts verilog2vhdl.sh and script.sed have been
  removed.

* Undocumented and ambiguous feature of the `spice-sdb' backend for
  gnetlist consisting in the possibility of using the "value="
  attribute instead of "file=" only for include blocks and only in
  embedding mode has been removed.

Notable changes in gEDA/gaf 1.9.1
=================================

* Unbalanced overbar markers are now rendered correctly.

* Most tools now no longer set the GEDADATA or GEDADATARC environment
  variables (they still obey them though).

* `gschem' now provides a path drawing tool, accessed using <A H> or
  the "Add Path" menu item.

  - Add a cusp node by left-clicking.

  - Add a Bezier curve node by left-dragging from the node point to
    the control point.

  - End the path by clicking on the same node point twice, or clicking
    on the start node to close the path.

* On desktops that reparent menu bars (e.g. Ubuntu Unity) it is once
  again possible to paste the clipboard using the menus.

* A number of `gschem' editing behaviour changes were made:

  - Double-clicking on an entry in the "In Use" tab of the component
    selector now behaves the same way as double-clicking in the
    "Libraries" tab.

  - When new attributes are attached to a net, they are placed at a 50
    unit offset.

  - The text color is now used as the default color for `pinlabel='
    attributes.

  - Pin attributes are no longer automatically replaced when rotating
    or mirroring pins.

  - It is possible to edit a text object even when other non-text
    objects are selected.

  - Control grips are now always drawn at the same on-screen size.

* The multi-attribute editor dialog in `gschem' has been extended to
  support more cases:

  - It now supports editing attributes of multiple selected
    objects. These are aggregated into the dialog's attribute list.

    Multiple attributes with the same name attached to any given
    object are still shown in separate rows, but identically named
    attributes from multiple selected objects will share a row. If a
    row has differing values between objects, its value shows the text
    "<various>" in a muted colour.

  - Where an attribute is present in some, but not all selected
    objects its name in the list is shown in red. A new popup menu
    option allows copying such attributes to all selected objects.

  - If no complex, placeholder, pin, net or bus objects are selected,
    the dialog now falls back to editing any selected attribute text.

    This allows editing of unattached or floating attributes in
    symbols, or any attributes seleccted independantly of their owning
    object. In this fallback mode, all attributes with the same name
    shown in separate rows, and the "add attribute" functionality is
    disabled.

* `gschem' has new icons for many editing commands.

* Several bugs were fixed in `gnetlist':

  - Invalid attributes are ignored.

  - Correct detection of graphical symbols.

  - Components with no `refdes=' attribute are reported.

  - Crash fixes for `drc' and `switcap' backends.

* `gaf export' now accepts a `--scale' option.

* The `--size', `--margins' and `--align' options to `gaf export' can
  now accept their multiple arguments separated with a space ` ', a
  colon `:' or a semicolon `;'.

* `gaf config' now handles the `-p' option properly.

* `gnetlist', `gsymcheck' and `gaf' now output internationalised
  messages.

* New Scheme procedures were added. Please see the "geda-scheme" info
  manual for more details.

  - The (gschem action) module contains procedures for working with
    user editing actions (i.e. commands accessed via the menus or
    keybindings).

  - Two new hooks were added: `bind-keys-hook' is called whenever
    keymap bindings are changed; `action-property-hook' is called when
    action properties are changed.


Notable changes in gEDA/gaf 1.9.0
=================================

* A new schematic renderer library, `libgedacairo' has been added, and
  all tools have been updated to use it.  Related changes:

  - The following configuration functions have been removed:

         output-capstyle         paper-sizes
         output-color            postscript-prolog
         output-orientation      print-command
         output-type             setpagedevice-orientation
         paper-size              setpagedevice-pagesize

  - The following `gschem' configuration items have been added:

         print-paper             print-color
         print-orientation

  - `gschem' now generates PDF files rather than PostScript or EPS
    files.

  - `gschem' now uses a standard GTK print dialog (with printer
    selection, print preview, etc.)

* A new `gaf' command-line utility has been added.  It provides a
  number of subcommands:

  - `gaf export' can be used to convert schematic files to PNG, SVG,
    PDF, PS and EPS, with detailed control of paper size and margins.
    It supports multi-page PDF and PS output.

  - `gaf config' can be used to manipulate the project, user, and
    system configuration stores (see below).

  - `gaf shell' provides a Scheme REPL for batch manipulation of
    schematic and symbol files.

* New project, user and system configuration stores have been added,
  using a format similar to .desktop files.  They can be manipulated
  using the `gaf config' tool.  Configuration parameters for most
  tools will be converted to use this mechanism in future releases.

* The changes to highlighting of dangling net ends in gEDA/gaf 1.8.0

utils/scripts/gnet_hier_verilog.sh
#!/bin/bash
# gEDA - GPL Electronic Design Automation
# Copyright (C) 2007-2008 Paul Tan (pt75234 at users.sourceforge.net)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02111-1301 USA.
# ========================================================
# Description:  
#   Generate non-flatten hierarchical Verilog netlist
# gnet_hier_verilog.sh
# version 0.0.2
# Usage: 
#   [path]gnet_hier_verilog.sh [path]FileName.sch
# Requires gawk and gEDA's gnetlist gnet-verilog.scm
# Please set tab = 2 for readability
# Written by Paul Tan
# ========================================================
# 1) This is a simple draft bash script to produce a
#    hierarchical verilog netlist in a single file. 
#    It gathers hierarchical information from a list of 
#    unique symbols/schematics originating from the top level
#    schematic all the way down to the lowest level of the 
#    design hierarchy. It then successively invokes the
#    existing gEDA verilog netlister to produce each single 
#    level netlists, and concatinates all the unique 
#    module netlists into one single hierarchical netlist 
#    file. 
# 2) Currently, it assumes that one or more hierarchical symbol
#    can be represented by a single schematic file. If needed, 
#    feature for mutiple schematic files mapped to a single symbol 
#    can be easily added. In that case, multiple source attrib
#    are used in that symbol.
# 3) It checks the follwoing errors while traversing down 
#    the hierarchy and terminates the netlisting if error 
#    is found.
#    a) if a symbol's source attribute indicated schematic can not
#       be found in the search paths defined by gafrc.
#    b) if a symbol's device attribute value does not match its
#       corresponding schematic's module_name attribute value.
# 4) This script assumes that there are no other errors in the 
#    entire hierarchy, and that the user has already run the
#    DRC. Moreover, it assumes that the user has run the single
#    level verilog netlister on each schematic in the hierarchy
#    without any error. it also assumes that the hierarchy-traverse
#    is disabled in gnetlistrc, thus disables flatten hierarchical 
#    netlist generation.
# 5) Netlist of modules are listed from top down, can easily be 
#    changed to do bottom up.
# 6) Symbol must contain "source=????.sch" attribute to have its
#    schematic netlisted. Otherwise, symbol is treated just as 
#    primitive instances in the netlist.
# 7) It only uses the gafrc file in the folder where the top level 
#    schematic resides. The search path for the symbols and
#    schematics should be defined in that gafrc file. The 
#    current implementation searches from the beginning
#    of the file, it will be changed to conform to the gEDA
#    practice of searching from the bottom first.
# 8) Hierarchy info is output to a report file for reference.
# 9)This script is optimized for readability only, I hope.
# 10)Array variables are suffixed with A, e,g,, HLineA[]
# ========================================================
#             To do list
# -------------------------------------------------------
# 1) Multi-page sch per symbol support
# 2) Use the "file" attrib to include user defined netlist
# 3) Include .gnetlistrc and other config files info to search
# 4) Change search order for sym/sch libs to LIFO (last in fisrt out)
# 5) Recode the script in C or scheme ?
# 6) Add CL args to support VHDL and Spice hier netlist
# 7) Add error checks:
#     a) add ERC (Electrical_Rule_Check) if not already checked
#        by gnetlist DRC.
# ========================================================
#
#                   Notes
#
# ========================================================
# First we create a master list in a record array HierListA[].
# HierListA[] will contain a list of records of all unique symbols
# under the top schematics and all its underlying schematics.
# This master record list will be output to a text file
# for reference.
# Then we'll invoke the gnetlist program to netlist each of 
# the schematics representing the symbols on the list,
# concatinating each netlist into the top level netlist.
#
# Each record of HierListA[] has 5 or more fields:
#  0          1         2          3         4        [5] ...
#  ItemNum    Flag     Sym_Name   Sch_Dir  Sch_Name  [Sch_Name] ..
# -------------------------------------------------------
# Corresponding variables declared in this script
# $iN         $iFlag    $iSym     $iSchDir $iSch
# $CurIndex   $CurFlag  $CurSym   $CurDir  $CurSch   ; In master
# $EndIndex   $Tflag    $Tsym     $TschDir $Tsch     ; To add
# -------------------------------------------------------
# The Flag is used to indicate if a schematic needs to 
# be netlisted.
#    Flag = 0 means NOT to netlist the schematic.
#    Flag = 1 means to netlist the schematic.
#    Flag = 2 means NOT to netlist, since other symbol already
#             has the same schematic listed.
# ----------------------------------------------
# For multiple pages(files) schematic, use more than 5 fields
# in a record.
# ========================================================
#
#          Script starts here
#
# ========================================================
# Configurable items
# ---------------------------------------
# Define the symbol's attribute to be used for its 
#   associated schematic.
src_attrib="source"
# ---------------------------------------
# version
ver=0.0.2
My_Debug=0
# ========================================================
my_error=0
my_ext=""
echo
if [ -z $1 ]
	then my_error=1
fi
# ---------------------------------------
# To be changed: loop thru multiple sch files from CL's args
if [ $my_error == 0 ]; then
	TopSchFile=$1
	my_ext=${TopSchFile#*.}
	if [ "$my_ext" != "sch" ]; then 
		my_error=1
	fi
fi
# ---------------------------------------
if [ $my_error == 1 ]; then
	echo "Error:"
	echo -n "Please include a gEDA schematic file"
	echo " as its argument."
	echo " Usage:"
	echo "   [path]gnet_hier_verilog.sh [path]FileName.sch"
	exit 1
fi
# ---------------------------------------
TopDir=$(dirname $1)
cd $TopDir
TopDir=$PWD
TopBaseFile=$(basename $1 .sch)
CurSchFile=$1
TopVFile="${TopBaseFile}.v"
ReportFile="${TopBaseFile}_Report.txt"
TemptFile="hier_temp.v"
# ---------------------------------------
echo
echo "Starting Hierarchical Verilog Netlister version ${ver} ....."
# ===============================================
if [ $My_Debug == 1 ]
	then
		echo 
		echo '======== Start Debug ==========='
		echo "TopDir = $TopDir"
		echo "TopBaseFile = $TopBaseFile"
		echo "CurSchFile = $CurSchFile"
		echo "ReportFile = $ReportFile"
fi
# ----------------------------------------------
#
# Init all indices:
# Fields position in a record
iN=0; iFlag=1; iSym=2; iSchDir=3; iSch=4
# ----------------------------------------------
# Initialize master list HierListA[]'s first record
CurIndex=0
EndIndex=0 
Tflag=1
Tsym="${TopBaseFile}.sym" 
Tsch="${TopBaseFile}.sch" 
TschDir="."
# Fields in each row of record
HLineA[$iN]=$EndIndex
HLineA[$iSym]=$Tsym
HLineA[$iSch]=$Tsch
HLineA[$iSchDir]=$TschDir
HLineA[$iFlag]=$Tflag
# First row of record in master list
HierListA[0]=`echo ${HLineA[@]:0}`
# ----------------------------------------------
# Initialize other lists
# ----------------------------------------------
# Get gEDA gafrc's symbol and schematic search paths
# into their respective list arrays
SymDirListA=( `
	gawk 'BEGIN {FS="\""}
		(NF >= 1)&&($1 == "(component-library ") {
		printf("%s\n",$2);
	}' gafrc` )
# ----------------------------------------------
SrcDirListA=( `
	gawk 'BEGIN {FS="\""}
		(NF >= 1)&&($1 == "(source-library ") {
		printf("%s\n",$2);
	}' gafrc` )

# ===============================================
if [ $My_Debug == 1 ]; then
	echo "---------------------------"
	echo -n "SymDirList n = "
	echo ${#SymDirListA[*]}
	for i in "${SymDirListA[@]}"; do echo "$i"; done
	echo "---------------------------"
	echo -n "SrcDirList n = "
	echo ${#SrcDirListA[*]}
	for i in "${SrcDirListA[@]}"; do echo "$i"; done
	echo "---------------------------"
fi
# ###############################################
# Process each record from the master list
while (("$CurIndex" <= "$EndIndex"))
do
	# Get a record from the master list
	CurA=( ${HierListA[$CurIndex]} )
	CurFlag=${CurA[$iFlag]}
	CurSchFile=${CurA[$iSch]}
	CurSchDir=${CurA[$iSchDir]}
	if [ "$CurSchDir" == "." ] 
		then CurSchPath=$CurSchFile
		else CurSchPath=`echo "$CurSchDir/$CurSchFile"`
	fi
	CurSchBase=$(basename $CurSchFile .sch)
	# --------------------------------
	# Only Flag=1 entry will get netlisted and processed
	# down the hierarchy
	if [ "$CurFlag" != 1 ]; then
		let "++CurIndex"
		continue
	fi
	# ----------------------------------------------
	# if needed, this is the place to generate netlist
	#	for each record. For now, we netlist at the end
	# after we have generated all the records
	# ----------------------------------------------
	# To be changed: loop thru pages of schs
	# ----------------------------------------------
	# Get all instances of any symbols from current schematic
	# into TempSymListA[]
	TempSymListA=( `
		gawk '(NF==7)&&($1=="C") {printf("%s\n",$7)}' $CurSchPath 
	`)
	# ----------------------------------------------
	# Make a unique symbol list from it
	TSymListA[0]=${TempSymListA[0]}
	found=0
	let "k = 0"
	for i in "${TempSymListA[@]}"; do
		found=0
		for j in "${TSymListA[@]}"; do
			if [ "$i" == "$j" ]; then 
				found=1
				break
			fi
		done # for j
		if [ $found == 1 ]; then continue
		fi
		let "++k"
		TSymListA[$k]=$i
	done # for i
	unset TempSymListA
	# ===============================================
	# Check if Tsym already exist in the master list
	# if not on list, add to the list
	found=0
	for Tsym in "${TSymListA[@]}"; do
		found=0
		for HLineA in "${HierListA[@]}"; do
			HierA=( `echo $HLineA` )
			HierSym=${HierA[$iSym]}

			if [ "$Tsym" == "$HierSym" ]; then
				found=1
				break
			fi
		done

		if [ $found == 1 ]; then continue
		fi
		# ===================================
		# Find the Symbol file and its folder
		found=0
		for i in "${SymDirListA[@]}"; do
			TsymPath=`echo "$i/$Tsym"`
			if [ -e $TsymPath ]; then
				found=1
				TsymDir=$i
				break
			fi 
		done

		if [ $found == 0 ]; then
			echo "Error: "
			echo "  $Tsym file not found. Please check the search path"
			echo "  defined in the gafrc file."
			exit 1
		fi
		# ---------------------------------
		# Find the schematic file name for the symbol from:
		#   1) The "source" attribute in the symbol file of 
		#      the $TsymDir folder.
		#   2) If no "source", use the symbol basename as the schematic
		#      basename?  For now, must use "source". 
		#   3) Note that the sch file must contain "module_name" 
		#      attrubute whose value must match the value of the 
		#      "device" attribute from the sym file. This script
		#      checks for this condition.
		# Given the schematic file name, search its folder path in:
		#   1) Source folders specified in the gEDA rc file
		# ---------------------------------
		TsymBase=$(basename $Tsym .sym)
		TschDir=$TsymDir
		Tflag=0
		# ---------------------------------
		# Get "source" and "device" attributes from sym file
		# and store them in SymAttribListA[]
		SymAttribListA=( `
			gawk -v src=$src_attrib 'BEGIN {FS="="}
				(NF==2)&&($1 == src) {print $0};
				(NF==2)&&($1 == "device") {print $0}
			' $TsymPath` )
		# ---------------------------------
		# To be changed: add a list of sources SymSrcList[]
		#                to support multi-page
		# ---------------------------------
		# Check if source exist in SymAttribListA[]
		Tsch=""
		for i in "${SymAttribListA[@]}"; do
			if [ "${i%=*}" == "$src_attrib" ]; then
				Tsch=${i#*=}; break
			fi
		done
		# ---------------------------------
		if [ "$Tsch" == "" ]; then 
				Tsch="None"; Tflag=0
			else
				# ---------------------------------
				# Found sch name, check if already on master list
				found=0
				for HLineA in "${HierListA[@]}"; do
					HierA=( `echo $HLineA` )
					HierSch=${HierA[$iSch]}
					if [ "$Tsch" == "$HierSch" ]; then	
						found=1
						break
					fi
				done
				# If sch not on master list, need further check
				if [ $found == 0 ]; then
					Tflag=2
				fi
		fi
		# ---------------------------------
		# if sch not on list, search for sch path
		if [ $Tflag == 2 ]; then
			found=0		
			for i in "${SrcDirListA[@]}"; do
				TschPath=`echo "$i/$Tsch"`
				if [ -e $TschPath ]; then
					found=1
					TschDir=$i
					break
				fi 
			done
			# if can't find sch path, error and exit
			if [ $found == 0 ]; then
				echo -n "Error: $Tsch is the $src_attrib attribute in "
				echo -n "$Tsym but NOT in the gafrc "
				echo "source-directory path."
				Tflag=3
				exit 1
			fi
			# ---------------------------------
			# Found sch path, check if sch's
			# sch's module_name attribute = sym's device attribute
			# ---------------------------------
			# Get module_name attrib from sch file
			TModule=""
			TModule=( `
				gawk 'BEGIN {FS="="}
					(NF == 2)&&($1 == "module_name") {
						printf("%s\n", $2); exit}' $TschPath` )
			# ---------------------------------
			# Check if module_name = device from SymAttribListA[]
			# ---------------------------------	
			Tdev=""		
			for i in "${SymAttribListA[@]}"; do
				if [ "${i%=*}" == "device" ]; then
					Tdev=${i#*=}; break
				fi
			done
			# ---------------------------------
			if [[ "$TModule" != "" ]] && [[ "$TModule" == "$Tdev" ]]
				then Tflag=1
				else 
					echo "Error: "
					echo "  $Tsym Symbol's device attribute $Tdev does not match"
					echo "  $Tsch schematic's module_name attribute $TModule."
				exit 1
			fi
			# ---------------------------------
		fi
		# ===================================
		# Add a new record into the master list HierListA[]
		let "++EndIndex"
		HLineA[$iN]=$EndIndex
		HLineA[$iFlag]=$Tflag
		HLineA[$iSym]=$Tsym
		HLineA[$iSchDir]=$TschDir
		HLineA[$iSch]=$Tsch
		HierListA[$EndIndex]=`echo ${HLineA[@]:0}`
		# ---------------------------------
		if [ $My_Debug == 3 ]; then
			echo "Added new record"
			echo ${HierListA[$EndIndex]}
			echo
		fi
		# ---------------------------------
	done	# done. for Tsym
	unset TSymListA
	# ===============================================
	let "++CurIndex"
done	# done generating master list
#
# ###############################################
# Generate a report file
echo "# Hierarchical netlist report for:" >$ReportFile
echo "# $TopSchFile" >>$ReportFile
echo "# ------------------------">>$ReportFile
# comments fields
HLineA[$iN]="Index"
HLineA[$iFlag]="Flag"
HLineA[$iSym]="Sym"
HLineA[$iSch]="Sch"
HLineA[$iSchDir]="Sch_Dir"
echo "# ${HLineA[@]:0}" >>$ReportFile
echo "# ------------------------">>$ReportFile
for i in "${HierListA[@]}"; do
	echo $i >>$ReportFile
done

# ###############################################
# Generate netlist from record in the master list
# Only Flag=1 record will be netlisted
# ###############################################
echo
echo "Generating hierarchical netlist ...... "
let "CurIndex=0"
let "EndIndex=${#HierListA[@]}-1"
# -----------------------------------
if [ $My_Debug == 2 ]; then
	echo
	echo "CurIndex = $CurIndex"
	echo "EndIndex = $EndIndex"
	echo
fi
# -----------------------------------
while (("$CurIndex" <= "$EndIndex"))
do
	# Get a record from the master list
	CurA=( ${HierListA[$CurIndex]} )
	# Get fields of the selected record
	CurFlag=${CurA[$iFlag]}
	CurSchFile=${CurA[$iSch]}
	CurSchDir=${CurA[$iSchDir]}
	if [ "$CurSchDir" == "." ]
		then CurSchPath=$CurSchFile
		else CurSchPath=`echo "$CurSchDir/$CurSchFile"`
	fi
	CurSchBase=$(basename $CurSchFile .sch)
	# --------------------------------
	# Only Flag=1 entry will get netlisted
	if [ $CurFlag != 1 ]; then
		let "++CurIndex"
		continue
	fi
	# --------------------------------
	# To be changed: add multi-page netlist
	# --------------------------------
	if [ "$CurIndex" -eq 0 ]
		then 
			gnetlist -g verilog -o $TopVFile $TopSchFile
		else
			gnetlist -g verilog -o $TemptFile $CurSchPath
			# Skip header portion of each netlist
			gawk 'BEGIN{found = 0}
				(NF >= 1)&&($1 == "module")&&(found == 0){
					found = 1;
					printf("\n/* --------------------------------- */\n");
				};
				(NR > 0)&&(found == 1){
					print $0;
				};
			' $TemptFile >>$TopVFile
	fi
	# --------------------------------
	let "++CurIndex"
done
echo
echo "Hierarchical Verilog netlist successfully completed."
echo

 	  	 
